Ignore query params when using customRoutes with SocketModeReceiver

moustacheful opened the pull request:
###  Summary

https://github.com/slackapi/bolt-js/issues/1206

Allows for using query params in custom routes. Currently it will attempt to match the whole url including the query params.
With this change, it will skip the query params when matching.

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slackapi/bolt-js?pullRequest=1207) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slackapi/bolt-js?pullRequest=1207) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slackapi/bolt-js?pullRequest=1207) it.</sub>

misscoded commented:
Thanks again for taking the initiative to address the bug you've identified, @moustacheful! 🙌🏼 

The same (flawed) logic is also present in the `HTTPReceiver`. If possible, would you mind implementing the same fix there, as well as the corresponding test?

moustacheful commented:
> Thanks again for taking the initiative to address the bug you've identified, @moustacheful! 🙌🏼
> 
> The same (flawed) logic is also present in the `HTTPReceiver`. If possible, would you mind implementing the same fix there, as well as the corresponding test?

Hey! no problem. I believe the HTTPReceiver might not run into the same issue because of this:
https://github.com/slackapi/bolt-js/blob/main/src/receivers/HTTPReceiver.ts#L436

[URL.pathname](https://nodejs.org/api/url.html#urlpathname) does not include the query params section of the URL.

that is, unless I'm looking at the wrong file :detective:

moustacheful commented:
I've updated the PR to have the same implementation as HTTPReceiver, using URL instead. I'm sure it can save us some headache on some random edge case I might not have been considering with the `.split('?')`.

misscoded commented on src/receivers/SocketModeReceiver.ts:
```diff
@@ -187,10 +188,13 @@ export default class SocketModeReceiver implements Receiver {
-          const match = this.routes[req.url] && this.routes[req.url][method] !== undefined;
+          // NOTE: the domain and scheme are irrelevant here.
+          // The URL object is only used to safely obtain the path to match
+          const { pathname: path } = new URL(req.url as string, 'http://localhost');
```
We probably want to switch this hard-coded string out for the same `http://${req.headers.host}` that's featured in the `HTTPReceiver` implementation.

I believe that as a result, one of the tests will fail and the `fakeReq` will need to be augmented with the `headers` object (similar to what's seen in the test you've introduced).

moustacheful replied:
I actually changed it from that because it made it look like it had a purpose. As the comment on the other side explained

```
    // NOTE: the domain and scheme of the following URL object are not necessarily accurate. The URL object is only
    // meant to be used to parse the path and query
```

it doesn't actually serve any purpose beyond extracting the pathname, it's just there because URL _needs_ a URL and not a URI. 

I don't feel too strongly though, let me know if you still want it that way, I can fix it tomorrow.

seratch replied:
@moustacheful Thanks for elaborating here! 

You're right that, for the second argument in the `URL` constructor, choosing either `localhost` or `req.headers.host` does not matter in this scenario. That being said, we don't have a strong reason to use `localhost` (I know that's a common code example, though). Plus, we prefer consistency across the code base in general. 

Could you do the same with the code in `HTTPReceiver` tomorrow? Thanks again for taking the time to report and fix this issue!

moustacheful replied:
Hey yeah, no problem, I've updated `HTTPReceiver` to be consistent with `SocketReceiver`, and added a test case for this scenario on the former.

misscoded reviewed (changes requested):
You're absolutely right! I'd entirely forgotten that the two receivers handled the `req.url` differently.

I left a note about the hardcoded `localhost` reference and what I anticipate will be an associated failing test, but other than that, everything looks to be in order. 👍🏼

filmaj reviewed (approved):
Looks good to me, thanks for creating the PR!

seratch commented:
My suggestion in my last comment was to use `req.headers.host` for both receivers but the changes here are the opposite (using localhost consistently). 

That being said, how we construct the URL objects is not important here. Either way, we don't extract the hostname value from URL objects. If other maintainers are fine with this change, please go ahead 👍

misscoded reviewed (approved):
Since the other maintainers have expressed that they're OK with this change, I won't block it
